usr/share/rear/prep/RSYNC/default/150_check_rsync_protocol_version.sh

```diff
@@ -42,7 +42,7 @@ if [ "${RSYNC_USER}" != "root" ]; then
             StopIfError "Remote file system $_mntpt does not have user_xattr mount option set!"
             #BACKUP_RSYNC_OPTIONS+=( --xattrs --rsync-path="""rsync --fake-super""" )
             # see issue #366 for explanation of removing --xattrs
-            BACKUP_RSYNC_OPTIONS+=( --rsync-path="""rsync --fake-super""" )
+            BACKUP_RSYNC_OPTIONS+=( --rsync-path=\"rsync --fake-super\" )
```

Comment:
I am not a BACKUP=RSYNC user
but from plain looking at the code I wonder
if single quotes wouldn't be even simpler as in
```
BACKUP_RSYNC_OPTIONS+=( --rsync-path='rsync --fake-super' )
```
because - as far as I see - the `rsync --fake-super` value
seems to be a constant string so single quotes quoting should work?

Reply:
OK - I got it - seems single quotes are wrong because
```
# BACKUP_RSYNC_OPTIONS=( foo bar )

# BACKUP_RSYNC_OPTIONS+=( --rsync-path='rsync --fake-super' )

# for e in "${BACKUP_RSYNC_OPTIONS[@]}" ; do echo "'$e'" ; done
'foo'
'bar'
'--rsync-path=rsync --fake-super'
```
but also `\"...\"` looks wrong to me
```
# unset BACKUP_RSYNC_OPTIONS

# BACKUP_RSYNC_OPTIONS=( foo bar )

# BACKUP_RSYNC_OPTIONS+=( --rsync-path=\"rsync --fake-super\" )

# for e in "${BACKUP_RSYNC_OPTIONS[@]}" ; do echo "'$e'" ; done
'foo'
'bar'
'--rsync-path="rsync'
'--fake-super"'
```
while no quoting at all looks right to me
```
# unset BACKUP_RSYNC_OPTIONS

# BACKUP_RSYNC_OPTIONS=( foo bar )

# BACKUP_RSYNC_OPTIONS+=( --rsync-path=rsync --fake-super )

# for e in "${BACKUP_RSYNC_OPTIONS[@]}" ; do echo "'$e'" ; done
'foo'
'bar'
'--rsync-path=rsync'
'--fake-super'
```

But I may misunderstand things here
because I am not a BACKUP=RSYNC user
so I don't know how things actually behave in practice.

Reply:
Hello,

The result must be:
rsync ... --rsync-path="rsync --fake-super" ...

- A workaround with a recent rsync is: -M--fake-super
Thanks

Reply:
When the result should be `--rsync-path="rsync --fake-super"`
then I think
```
BACKUP_RSYNC_OPTIONS+=( '--rsync-path="rsync --fake-super"' )
```
would be the most clear and straightforward way because
```
# unset BACKUP_RSYNC_OPTIONS

# BACKUP_RSYNC_OPTIONS=( foo bar )

# BACKUP_RSYNC_OPTIONS+=( '--rsync-path="rsync --fake-super"' )

# for e in "${BACKUP_RSYNC_OPTIONS[@]}" ; do echo "$e" ; done
foo
bar
--rsync-path="rsync --fake-super"
```

In contrast
```
BACKUP_RSYNC_OPTIONS+=( --rsync-path=\"rsync --fake-super\" )
```
only works when the array elements are evaluated via
`${BACKUP_RSYNC_OPTIONS[*]}` or
`${BACKUP_RSYNC_OPTIONS[@]}` or
`"${BACKUP_RSYNC_OPTIONS[@]}"`
because
```
# unset BACKUP_RSYNC_OPTIONS

# BACKUP_RSYNC_OPTIONS=( foo bar )

# BACKUP_RSYNC_OPTIONS+=( --rsync-path=\"rsync --fake-super\" )

# echo ${BACKUP_RSYNC_OPTIONS[*]}
foo bar --rsync-path="rsync --fake-super"

# echo ${BACKUP_RSYNC_OPTIONS[@]}
foo bar --rsync-path="rsync --fake-super"

echo "${BACKUP_RSYNC_OPTIONS[@]}"
foo bar --rsync-path="rsync --fake-super"
```

As far as I see in current code the BACKUP_RSYNC_OPTIONS array
is evaluated always via `BACKUP_RSYNC_OPTIONS[@]` so
```
BACKUP_RSYNC_OPTIONS+=( --rsync-path=\"rsync --fake-super\" )
```
would work with current code but I think
```
BACKUP_RSYNC_OPTIONS+=( '--rsync-path="rsync --fake-super"' )
```
would be more fail safe code because it would allow
to also evaluate the BACKUP_RSYNC_OPTIONS array
element by element.

Reply:
Thanks !

I'm working on the RHEL version, thus I was not impacted by the typo of BACKUP_RSUNC_OPTIONS.

I tried BACKUP_RSYNC_OPTIONS+=( '--rsync-path="rsync --fake-super"' )
but even if the the command appears ok from the log:
`
++ echo '2021-03-02 11:39:07.015580343 rsync --sparse --archive --hard-links --numeric-ids --stats --rsync-path="rsync --fake-super" --relative --one-file-system --delete --exclude-from=/tmp/rear.7zjuowy7jEiOXBA/tmp/backup-exclude.txt --delete-excluded / /boot/efi /boot user1@192.168.122.220:/home/user1/rear//localhost/backup'
2021-03-02 11:39:07.015580343 rsync --sparse --archive --hard-links --numeric-ids --stats --rsync-path="rsync --fake-super" --relative --one-file-system --delete --exclude-from=/tmp/rear.7zjuowy7jEiOXBA/tmp/backup-exclude.txt --delete-excluded / /boot/efi /boot user1@192.168.122.220:/home/user1/rear//localhost/backup
+++ cat /tmp/rear.7zjuowy7jEiOXBA/tmp/backup-include.txt
`

- it is failing: 
`
++ rsync --sparse --archive --hard-links --numeric-ids --stats '--rsync-path="rsync' '--fake-super"' --relative --one-file-system --delete --exclude-from=/tmp/rear.7zjuowy7jEiOXBA/tmp/backup-exclude.txt --delete-excluded / /boot/efi /boot user1@192.168.122.220:/home/user1/rear//localhost/backup
rsync: --fake-super": unknown option
rsync error: syntax or usage error (code 1) at main.c(1578) [client=3.1.3]
`
- And "BACKUP_RSYNC_OPTIONS+=( --rsync-path=\"rsync --fake-super\" )" is also failing ....

- The best solution seems to be "-M--fake-super", thus no need of ""

Reply:
@Reiner030 
thank you for your feedback.
It helps so much because it makes the actual problem obvious.

I will investigate how to implement it correctly in general
to specify options with quotation marks as array elements
and how to pass them properly to commands.

Reply:
@bwelterl
I need the complete ReaR debug log file from you
where you posted the excerpts in your
https://github.com/rear/rear/pull/2577#discussion_r585740764

Reply:
Please find the log with BACKUP_RSYNC_OPTIONS+=( '--rsync-path="rsync --fake-super"' )
(sorry or the delay)

[rear-localhost.log](https://github.com/rear/rear/files/6150375/rear-localhost.log)

thank you

Reply:
The original quoting was correct, although needlessly complicated:
`--rsync-path="""rsync --fake-super"""`
is exactly equivalent to
`--rsync-path="rsync --fake-super"`
or even
`"--rsync-path=rsync --fake-super"`
@bwelterl @jsmeix you seem to think that it is necessary to get quotes inside the string (two levels of quoting) somehow, but that does not seem to be true. One level of quoting (around the whole string with whitespace to prevent splitting on whitespace) is perfectly enough. The argument passed to rsync does not need to be quoted: rsync does not want to see the quotes and the shell calling rsync does not perform quote removal after expanding the `BACKUP_RSYNC_OPTIONS` (because `${BACKUP_RSYNC_OPTIONS[@]}` is itself expanded inside quotes in 500_make_rsync_backup.sh :
https://github.com/rear/rear/blob/3a6993968852791a63fac49bce6e81d52e258622/usr/share/rear/backup/RSYNC/default/500_make_rsync_backup.sh#L27 so you certainly don't want to have quotes inside the string, because they would get passed to the rsync command and confuse everything, i.e. you would get errors like `rsync --fake-super`: command not found, because the whole string would be considered a command path.)

If you wonder why the code does not work then when it is correct, it is probably because this code destroys whitespace in the array: https://github.com/rear/rear/blob/aa7b1979c347a3800959d757f7ce485fc10bd788/usr/share/rear/backup/RSYNC/default/200_check_rsync_relative_option.sh#L8

It is now fixed by @jsmeix using `+=` - @bwelterl I have verified that the current development version does not suffer from the problem anymore (I was able to reproduce the problem in 2.4 but not in 2.6), so I will close the issue.

Reply:
> OK - I got it - seems single quotes are wrong because
> 
> ```
> # BACKUP_RSYNC_OPTIONS=( foo bar )
> 
> # BACKUP_RSYNC_OPTIONS+=( --rsync-path='rsync --fake-super' )
> 
> # for e in "${BACKUP_RSYNC_OPTIONS[@]}" ; do echo "'$e'" ; done
> 'foo'
> 'bar'
> '--rsync-path=rsync --fake-super'
> ```

This is, actually, exactly the expected result. You want the `rsync --fake-super` to be one word, no splitting.

So yes, plain simple quotes are enough. When I touch the code, I will convert the double triple double quotes to a pair of simple quotes.

> while no quoting at all looks right to me
> 
> ```
> # unset BACKUP_RSYNC_OPTIONS
> 
> # BACKUP_RSYNC_OPTIONS=( foo bar )
> 
> # BACKUP_RSYNC_OPTIONS+=( --rsync-path=rsync --fake-super )
> 
> # for e in "${BACKUP_RSYNC_OPTIONS[@]}" ; do echo "'$e'" ; done
> 'foo'
> 'bar'
> '--rsync-path=rsync'
> '--fake-super'
> ```

no, this is not correct. You need `rsync --fake-super` to be a single argument, without being split on space.